src/k8s_upgrade.py

```diff
@@ -0,0 +1,421 @@
+        """Handle upgrade events."""
+        if not self._upgrade:
+            logger.debug("Peer relation not available")
+            return
+        if not self._upgrade.versions_set:
+            logger.debug("Peer relation not ready")
+            return
+        if self.charm.unit.is_leader() and not self._upgrade.in_progress:
+            # Run before checking `self._upgrade.is_compatible` in case incompatible upgrade was
+            # forced & completed on all units.
+            self._upgrade.set_versions_in_app_databag()
+
+        if self._upgrade.unit_state is UnitState.RESTARTING:  # Kubernetes only
+            if not self._upgrade.is_compatible:
+                logger.info(
+                    "Refresh incompatible. If you accept potential *data loss* and *downtime*, you can continue with `force-refresh-start`"
+                )
+                self.charm.status.set_and_share_status(
+                    Config.Status.INCOMPATIBLE_UPGRADE
+                )
+                return
+        if (
+            not during_upgrade
+            and self.charm.db_initialised
+            and self.charm.is_db_service_ready()
+        ):
+            self._upgrade.unit_state = UnitState.HEALTHY
+            self.charm.status.set_and_share_status(ActiveStatus())
+        if self.charm.unit.is_leader():
+            self._upgrade.reconcile_partition()
+
+        self._set_upgrade_status()
+
+    def _set_upgrade_status(self):
+        if self.charm.unit.is_leader():
+            self.charm.app.status = self._upgrade.app_status or ActiveStatus()
+        # Set/clear upgrade unit status if no other unit status - upgrade status for units should
+        # have the lowest priority.
+        if (
+            isinstance(self.charm.unit.status, ActiveStatus)
+            or (
+                isinstance(self.charm.unit.status, BlockedStatus)
+                and self.charm.unit.status.message.startswith(
+                    "Rollback with `juju refresh`. Pre-refresh check failed:"
+                )
+            )
+            or self.charm.unit.status == Config.Status.WAITING_POST_UPGRADE_STATUS
+        ):
+            self.charm.status.set_and_share_status(
+                self._upgrade.get_unit_juju_status() or ActiveStatus()
+            )
+
+    def _on_upgrade_peer_relation_created(self, _) -> None:
+        if self.charm.unit.is_leader():
+            self._upgrade.set_versions_in_app_databag()
+
+    def run_post_app_upgrade_task(self, event: EventBase):
+        """Runs post-upgrade checks for after mongos router upgrade."""
+        # The mongos service cannot be considered ready until it has a config-server. Therefore
+        # it is not necessary to do any sophisticated checks.
+        if not self.charm.mongos_initialised:
+            self._upgrade.unit_state = UnitState.HEALTHY
+            return
+
+        self.run_post_upgrade_checks(event)
+
+    def _on_pre_upgrade_check_action(self, event: ActionEvent) -> None:
```

Comment:
1. Is this action needed? I do not see it in the [mongos VM upgrades](https://github.com/canonical/mongos-operator/blob/6/edge/src/upgrades/mongos_upgrade.py)  
2. It doesn't seem like we're splitting the logic like we do in VM (`k8s_upgrade.py` and `mongos_upgrade.py`), can we stay consistent? or is there a specific reason we're not ?

Reply:
1. This was discussed with @delgod and he asked to have this action to be in line with k8s v3 upgrades
2. There's no reason, it's just a design choice. I can split it in two files to stay consistent yes!